File: vpon_plugin_poc/android/src/main/kotlin/io/flutter/plugins/vponmobileads/VponMobileAdsViewFactory.java
package io.flutter.plugins.vponmobileads;

import android.content.Context;
import android.graphics.Color;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.vpon_plugin_poc.BuildConfig;

import java.util.Locale;

import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;

class VponMobileAdsViewFactory extends PlatformViewFactory {

    @NonNull
    private final VponAdInstanceManager manager;

    VponMobileAdsViewFactory(@NonNull VponAdInstanceManager manager) {
        super(StandardMessageCodec.INSTANCE);
        this.manager = manager;
    }

    @NonNull
    @Override
    public PlatformView create(Context context, int viewId, @Nullable Object args) {
        if (args == null) {
            return getErrorView(context, 0);
        }
        final int adId = (Integer) args;
        VponFlutterAd ad = manager.adForId(adId);
        if (ad == null || ad.getPlatformView() == null) {
            return getErrorView(context, adId);
        }
        return ad.getPlatformView();
    }

    private static class ErrorTextView implements PlatformView {
        private final TextView textView;

        private ErrorTextView(Context context, String message) {
            textView = new TextView(context);
            textView.setText(message);
            textView.setBackgroundColor(Color.RED);
            textView.setTextColor(Color.YELLOW);
        }

        @Override
        public View getView() {
            return textView;
        }

        @Override
        public void dispose() {
        }
    }

    private static PlatformView getErrorView(@NonNull final Context context, int adId) {
        final String message =
                String.format(
                        Locale.getDefault(),
                        "This ad may have not been loaded or has been disposed. "
                                + "Ad with the following id could not be found: %d.",
                        adId);

        if (BuildConfig.DEBUG) {
            return new ErrorTextView(context, message);
        } else {
            Log.e(VponMobileAdsViewFactory.class.getSimpleName(), message);
            return new PlatformView() {
                @Override
                public View getView() {
                    return new View(context);
                }

                @Override
                public void dispose() {
                    // Do nothing.
                }
            };
        }
    }
}

File: vpon_plugin_poc/android/src/main/kotlin/io/flutter/plugins/vponmobileads/VponMobileAdsViewFactory.kt
package io.flutter.plugins.vponmobileads

import android.content.Context
import android.graphics.Color
import android.util.Log
import android.view.View
import android.widget.TextView
import com.example.vpon_plugin_poc.BuildConfig
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import java.util.Locale

internal class VponMobileAdsViewFactory(private val manager: VponAdInstanceManager) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        if (args == null) {
            return getErrorView(context!!, 0)
        }
        val adId = args as Int
        val ad: VponFlutterAd? = manager.adForId(adId)

        if (ad?.getPlatformView() == null) {
            return getErrorView(context!!, adId)
        }
        return ad.getPlatformView()!!
    }

    private class ErrorTextView constructor(context: Context, message: String) :
        PlatformView {

        private val textView = TextView(context)

        init {
            textView.setBackgroundColor(Color.RED)
            textView.setTextColor(Color.YELLOW)
            textView.text = message
        }

        override fun getView(): View {
            return textView
        }

        override fun dispose() {
        }

    }

    companion object {
        private fun getErrorView(context: Context, adId: Int): PlatformView {
            val message = String.format(
                Locale.getDefault(),
                "This ad may have not been loaded or has been disposed. "
                        + "Ad with the following id could not be found: %d.",
                adId
            )
            if (BuildConfig.DEBUG) {
                return ErrorTextView(context, message)
            } else {
                Log.e("VponAdsViewFactory", message)
                return object : PlatformView {
                    override fun getView(): View {
                        return View(context)
                    }

                    override fun dispose() {
                        // Do nothing.
                    }
                }
            }
        }
    }
}
